src/seutil/powershell.py

```diff
@@ -0,0 +1,141 @@
+            "0",
+            "false",
+            "False",
+        }
+        if show_full_output:
+            s += f"STDOUT:\n{self.stdout}\n"
+            s += f"STDERR:\n{self.stderr}\n"
+        else:
+            if len(self.stdout) > 800:
+                s += f"STDOUT (truncated):\n{self.stdout[:400]}...{self.stdout[-400:]}\n"
+            else:
+                s += f"STDOUT:\n{self.stdout}\n"
+            if len(self.stderr) > 800:
+                s += f"STDERR (truncated):\n{self.stderr[:400]}...{self.stderr[-400:]}\n"
+            else:
+                s += f"STDERR:\n{self.stderr}\n"
+        return s
+
+    def __repr__(self) -> str:
+        return self.__str__()
+
+
+def run(
+    cmd: str,
+    check_returncode: Optional[int] = None,
+    warn_nonzero: bool = True,
+    update_env: bool = False,
+    update_env_clear_existing: bool = False,
+    timeout: Optional[float] = None,
+    **kwargs,
+) -> subprocess.CompletedProcess:
+    """
+    Run a powershell command using subprocess.run.  The command will be run using "powershell -Command".
+
+    Some arguments' default values are changed (but can be overridden with kwargs):
+    * capture_output=True, text=True:  capture all stdout and stderr.
+
+    This function is able to check if return code match a given value (subprocess only supports
+    checking non-zero values, but this function supports any).  Nevertheless, this function
+    warns about any non-zero values if check_returncode is not set, to avoid silent failures;
+    this behavior can be turned off via warn_nonzero=False.
+
+    In addition, this function can try to update the environment variables in this process
+    with the ones after running the command (if the command finished successfully).
+    The retrieval of the sub shell's environments is done by `env` into a temporary file.
+
+    :param cmd: the command to run
+    :param check_returncode: the return code to expect from the command
+    :param warn_nonzero: whether to warn about non-zero exit codes
+    :param update_env: whether to update the environment variables in this process
+    :param update_env_clear_existing: whether to clear existing environment variables before updating
+    :param timeout: number of seconds to wait
+    :param kwargs: other arguments passed to subprocess.Popen
+    :return: the subprocess.CompletedProcess object, has stdout, stderr, returncode fields
+    :raises: PowerShellError if the command's output did not match check_returncode
+    :raises: subprocess.TimeoutExpired if the command timed out
+    """
+    # potentially append `env` to command to collect the environment variables
+    # TODO: this is hacky: it may mess up some commands; the env won't be collected when timeout; and variable values
+    # longer than 1 line will break the collection
+    if update_env:
+        tempfile_update_env = io.mktmp("seutil-powershell", ".txt")
+        cmd += f" ; env > {tempfile_update_env}"
```

Comment:
This part does not work in powershell for sure...

Reply:
I think this might work:
if update_env:
        tempfile_update_env = io.mktmp("seutil-powershell", ".txt")
        cmd += f" ; Get-ChildItem Env: > {tempfile_update_env}"